File: javascript/nodes/prompt_options.js
import { Next } from '../ambler.js';
import { createInterface } from 'readline';

const defaultUtils = {
    readLine: () => new Promise((resolve) => {
        const rl = createInterface({ input: process.stdin });
        rl.once('line', (line) => {
            rl.close();
            resolve(line.trim());
        });
    }),
};

export function promptOptions(edges, utils = defaultUtils) {
    return async (state) => {
        const hasKhinsiderUrls = state.urls.some((url) =>
            url.startsWith('https://downloads.khinsider.com/game-soundtracks')
        );

        const options = [
            ['quit', null],
            ['list', new Next(edges.onList, state)],
        ];

        if (hasKhinsiderUrls) {
            options.push(['resolve', new Next(edges.onResolve, state)]);
        } else {
            options.push(['download', new Next(edges.onDownload, state)]);
        }

        while (true) {
            console.log('\nSelect an option:');
            options.forEach(([name], i) => console.log(`${i + 1}. ${name}`));

            const line = await utils.readLine();
            const choice = parseInt(line, 10) - 1;
            if (choice >= 0 && choice < options.length) {
                return options[choice][1];
            }
            console.log('Invalid option. Please try again.');
        }
    };
}

File: ruby/nodes/prompt_options.rb
# frozen_string_literal: true

require_relative '../lib/ambler'

module PromptOptions
  DEFAULT_UTILS = { read_line: -> { $stdin.gets.to_s.chomp } }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    ->(state) {
      has_khinsider_urls = state[:urls].any? { |url|
        url.start_with?('https://downloads.khinsider.com/game-soundtracks')
      }

      options = [
        ['quit', nil],
        ['list', Ambler::Next.new(edges[:on_list], state)]
      ]

      if has_khinsider_urls
        options << ['resolve', Ambler::Next.new(edges[:on_resolve], state)]
      else
        options << ['download', Ambler::Next.new(edges[:on_download], state)]
      end

      loop do
        puts "\nSelect an option:"
        options.each_with_index { |(name, _), i| puts "#{i + 1}. #{name}" }

        line = utils[:read_line].call
        choice = Integer(line.strip, 10) - 1 rescue -1
        if choice >= 0 && choice < options.length
          return options[choice][1]
        end

        puts 'Invalid option. Please try again.'
      end
    }
  end
end
